master/buildbot/util/kubeclientservice.py

```diff
@@ -190,19 +198,37 @@ def __init__(self, kube_config=None):
         self.config = kube_config
         HTTPClientService.__init__(self, '')
         self._namespace = None
-        self.addService(kube_config)
+        kube_config.setServiceParent(self)
 
+    @defer.inlineCallbacks
     def _prepareRequest(self, ep, kwargs):
         config = self.config.getConfig()
         self._base_url = config['master_url']
         url, req_kwargs = HTTPClientService._prepareRequest(self, ep, kwargs)
 
+        if config['headers']:
+            if not req_kwargs['headers']:
+                req_kwargs['headers'] = {}
+            else:
+                req_kwargs['headers'] = req_kwargs['headers'].dup
+            req_kwargs['headers'].update(config['headers'])
+
+        bearer = config.get('bearerToken', None)
+        if bearer is not None:
+            bearer = yield self.config.renderSecrets(bearer)
+            req_kwargs['headers']['Authorization'] = "Bearer {token}".format(token=bearer)
+
+        basic = config.get('basicAuth', None)
+        if basic is not None:
+            basic = yield self.config.renderSecrets(basic)
```

Comment:
In python, usually, the basic auth conf is in the form ('login', 'password'). Secret rendering is recursive, so you can write

```python
KubeHardcodedConfig(basic=('user1', Secret('user1_passwd'))
```
So I think we should implement the ":".join() and the base64 encode.
And we should put that logic inside the base config loader class instead of here.

I would rather put getConfig a deferable, so that we can implement this logic, and also implement caching.

Reply:
> And we should put that logic inside the base config loader class instead of here.

By "base config loader", I'm assuming that you mean HardcodedConfig ?

Reply:
I was thinking about KubeConfigLoaderBase  https://github.com/buildbot/buildbot/blob/15826fcf9bc0e99d0a8640bf02a32911cdaf90fa/master/buildbot/util/kubeclientservice.py#L42

Reply:
You know best of course, but given that the authorization headers are specific to the hardcoded config class, I'm not so sure how to integrate them in the base config loader.

Anyways, moving things to getConfig is a can of worms, I've got to refactor the fake kube service to handle that :(

Reply:
right. The core reason is that I think _prepareRequest start to become too large, so maybe we can just split that into several methods a call it a day

Reply:
:+1: I'll do that instead.